### PR TITLE
Proper OpenID Connect auth implementation

### LINKED ISSUE #253: Refactor ORCID functionality into separate module

### LINKED ISSUE #322: As a user, I want to be able to log-in to SCope using ORCiD
Use the OpenID Connect Authorisation Code flow or the Hybrid flow through `FastAPI` to authenticate users.

### PR DESCRIPTION
**Pull request by MaybeJustJames**
This PR fleshes out an authentication/authorization mechanism using OpenID Connect [Authorization Code Flow](https://auth0.com/docs/flows/authorization-code-flow).

Further it begins to flesh out a deployment process by adding an administration command: `poetry run scope-console`. Using
this command an administrator can customize authentication providers, administrator user setup, and file upload limits. A
local OpenID Connect provider (KeyCloak) is used for local development using docker.

Closes #322, #253

To explore the API:
```bash
cd SCope/
docker-compose up -d && sleep 20 # The Keycloak server takes a while to start up
cd server/
./run-dev-env.sh
```
Then open a browser and point it to [http://localhost:8000/api/v1/docs](http://localhost:8000/api/v1/docs), or open [Postman](https://www.postman.com/) and import the project.

In order to get an access token:
1. Query the `/auth/loginurl` endpoint and copy the response URL into a browser
2. When prompted username is "scope" and password is "scope"
SCope login
3. From the redirect URL, copy the code parameter into the `code` field in the body of the `/auth/authorize` endpoint (The `state` field should already be filled properly if you're using Postman).
Copy code parameter
4. If you're using Postman to explore the API, the access token is automatically used by the other endpoints so you no longer need to worry about it.

### CODE REVIEW

File: server/scopeserver/oidc.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
+" Abstractions over OpenID Connect APIs "
```
Comment (dweemx) [inline server/scopeserver/oidc.py:1]: Maybe some links to docs would help here too

**Review by dweemx — COMMENTED** [review]
Overall very nice and quite exhaustive!
Maybe lacking some docs at certain places where code is not trivial (e.g.: `server/scopeserver/api/v1/auth.py`)

I tried the steps on this PR to use the OpenAPI docs but it doesn't work for me.

Authentication with ORCID on SCope is broken, I guess this will be fixed in another PR ?

File: README.md [hunk 2]
```diff
@@ -37,44 +39,42 @@ LOCAL_SCOPE_REPO="${HOME}/repos/SCope"
...
+```
 
-To enable colaborative annotations and login via ORCID ID, API credentials (`orcidAPIClientID`, `orcidAPIClientSecret` and `orcidAPIRedirectURI`) must be added to the config file provided.
-These can be generated at the [orcid developer tools page](https://orcid.org/developer-tools).
+5. Finally, run the server
 
-The `dataHashSecret` entry in the config file should be filled in with a randomly generated string for example from the python [secrets package](https://docs.python.org/3/library/secrets.html).
-This string will be used to salt all annotation data, allowing validation of data generated on the instance of SCope. Any changes in this string will invalidate all pre-existing annotations.
+```bash
+# TBD
```
[the remark this answers is not in the document]
Comment (MaybeJustJames) [reply README.md:76]: Not necessarily. Prod deploy story needs to be thought about more carefully

File: README.md [hunk 3]
```diff
@@ -37,44 +39,42 @@ LOCAL_SCOPE_REPO="${HOME}/repos/SCope"
 git clone https://github.com/aertslab/SCope "${LOCAL_SCOPE_REPO}"
 # Go to your local cloned SCope repository.
 cd "${LOCAL_SCOPE_REPO}"
-# Install SCope.
-npm install
+# Install server dependencies
+cd server/
+poetry install
```
[the remark this answers is not in the document]
Comment (MaybeJustJames) [reply README.md:44]: Yes this is old poetry

File: server/scopeserver/api/v1/projects.py [hunk 4]
```diff
@@ -12,121 +13,281 @@
...
-@router.get("/", summary="Get all projects accessible to the current user.", response_model=List[schemas.Project])
-async def my_projects(
-    db: Session = Depends(deps.get_db),
-    current_user: models.User = Depends(deps.get_current_user),
+@router.get("/all", summary="Get all projects", response_model=List[schemas.Project])
+async def all_projects(
+    database: Session = Depends(deps.get_db),
+    offset: int = 0,
+    limit: int = 100,
+    _admin: models.User = Depends(deps.get_current_admin),
```
[the remark this answers is not in the document]
Comment (MaybeJustJames) [reply server/scopeserver/api/v1/projects.py:22]: It's there to ensure only admins can access this API

File: server/scopeserver/api/v1/auth.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
+" API endpoints dealing with authentication and authorization. "
```
[the remark this answers is not in the document]
Comment (MaybeJustJames) [reply server/scopeserver/api/v1/auth.py:1]: Added to README

File: server/scopeserver/oidc.py (hunk 1 above)
Comment (MaybeJustJames) [reply server/scopeserver/oidc.py:1]: Added to README

File: server/scopeserver/api/v1/users.py [hunk 6]
```diff
@@ -1,17 +1,86 @@
...
+    offset: int = 0,
+    limit: int = 100,
+    _current_user: models.User = Depends(deps.get_current_admin),
+):
+    "Get a listing of all users"
+    return crud.get_users(database, offset, limit)
+
+
+@router.post("/promote", summary="Promote a user to an admin", response_model=schemas.UserResponse)
+async def promote(
```
Comment (KrisDavie) [inline server/scopeserver/api/v1/users.py:47]: Should this not require the submitting user to _be_ an admin to promote to admin? Or is this checked elsewhere?

File: server/scopeserver/api/v1/users.py [hunk 7]
```diff
@@ -1,17 +1,86 @@
...
+    database: Session = Depends(deps.get_db),
+    offset: int = 0,
+    limit: int = 100,
+    _current_user: models.User = Depends(deps.get_current_admin),
+):
+    "Get a listing of all users"
+    return crud.get_users(database, offset, limit)
+
+
+@router.post("/promote", summary="Promote a user to an admin", response_model=schemas.UserResponse)
```
Comment (KrisDavie) [inline server/scopeserver/api/v1/users.py:46]: Promote, demote and delete seem to be inconsistent in how they deal with checking if a user is admin before allowing them to run these.

File: server/test/api_fixture.py [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+    return httpx.Response(404, json={"error": "Unimplemented"})
+
+
+def mock_http_error_handler(request):
+    print(f"Got a mock request for error: {request.url.path}")
+    return httpx.Response(404, json={"error": "This is an intentional failure for testing purposes"})
+
+
+def mock_http_token_fails(request):
+    print(f"Got a mock request for bad-tonen mock: {request.url.path}")
```
Comment (KrisDavie) [inline server/test/api_fixture.py:70]:
```suggestion
    print(f"Got a mock request for bad-token: {request.url.path}")
```

File: server/test/api_fixture.py [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+                "token_endpoint": "http://localhost-token-error:8080/token",
+                "userinfo_endpoint": "http://localhost-token-error:8080/userinfo",
+                "authorization_endpoint": "http://localhost-token-error:8080/auth",
+            },
+        )
+    return httpx.Response(404, json={"error": "Test", "error_description": "Intentional test error message"})
+
+
+def mock_http_bad_key(request):
+    print(f"Got a mock request for bad-tonen mock: {request.url.path}")
```
Comment (KrisDavie) [inline server/test/api_fixture.py:86]:
```suggestion
    print(f"Got a mock request for bad-key: {request.url.path}")
```

**Review by KrisDavie — CHANGES_REQUESTED** [review]
Really nice work. Just a few comments/questioins

**Review by KrisDavie — APPROVED** [review]

**Pull request merged**